kvstore/pebble/kvstore.go

```diff
@@ -0,0 +1,163 @@
+func (store *pebbleKVStore) Get(key []byte) ([]byte, error) {
+	if key == nil {
+		return nil, ErrPebbleUnableToGetValue
+	}
+	value, closer, err := store.db.Get(key)
+	if err != nil {
+		if err == pebble.ErrNotFound {
+			return nil, ErrPebbleUnableToGetValue
+		}
+		return nil, errors.Join(ErrPebbleUnableToGetValue, err)
+	}
+	defer closer.Close()
+	return append([]byte{}, value...), nil
+}
+
+// Delete removes the key-value pair associated with the given key.
+func (store *pebbleKVStore) Delete(key []byte) error {
+	if key == nil {
+		return ErrPebbleUnableToDeleteValue
+	}
+	err := store.db.Delete(key, pebble.Sync)
+	if err != nil {
+		return errors.Join(ErrPebbleUnableToDeleteValue, err)
+	}
+	return nil
+}
+
+// GetAll retrieves all key-value pairs with keys starting with the given prefix.
+// If descending is true, it returns the results in reverse lexicographical order.
+func (store *pebbleKVStore) GetAll(prefix []byte, descending bool) (keys, values [][]byte, err error) {
+	iter, _ := store.db.NewIter(&pebble.IterOptions{
+		LowerBound: prefix,
+		UpperBound: prefixEndBytes(prefix),
+	})
+	defer iter.Close()
+
+	if descending {
+		for valid := iter.Last(); valid; valid = iter.Prev() {
+			keys = append(keys, append([]byte{}, iter.Key()...))
+			values = append(values, append([]byte{}, iter.Value()...))
+		}
+	} else {
+		for iter.First(); iter.Valid(); iter.Next() {
+			keys = append(keys, append([]byte{}, iter.Key()...))
+			values = append(values, append([]byte{}, iter.Value()...))
+		}
+	}
+
+	if err := iter.Error(); err != nil {
+		return nil, nil, errors.Join(ErrPebbleIteratingStore, err)
+	}
+
+	return keys, values, nil
+}
+
+// Exists checks if a key exists in the store and has a non-empty value.
+func (store *pebbleKVStore) Exists(key []byte) (bool, error) {
+	value, closer, err := store.db.Get(key)
+	if err == pebble.ErrNotFound {
+		return false, nil
+	}
+	if err != nil {
+		return false, errors.Join(ErrPebbleUnableToGetValue, err)
+	}
+	defer closer.Close()
+	return len(value) > 0, nil
+}
+
+// ClearAll removes all key-value pairs from the store.
+func (store *pebbleKVStore) ClearAll() error {
+	iter, _ := store.db.NewIter(nil)
+	defer iter.Close()
+	for iter.First(); iter.Valid(); iter.Next() {
+		if err := store.db.Delete(iter.Key(), pebble.Sync); err != nil {
+			return errors.Join(ErrPebbleClearingStore, err)
+		}
+	}
+	if err := iter.Error(); err != nil {
+		return errors.Join(ErrPebbleClearingStore, err)
+	}
+	return nil
+}
+
+// Stop closes the database connection.
+func (store *pebbleKVStore) Stop() error {
+	return store.db.Close()
+}
+
+// Len returns the number of key-value pairs in the store.
+func (store *pebbleKVStore) Len() (int, error) {
+	count := 0
+	iter, _ := store.db.NewIter(nil)
+	defer iter.Close()
+	for iter.First(); iter.Valid(); iter.Next() {
+		count++
+	}
+	if err := iter.Error(); err != nil {
+		return 0, errors.Join(ErrPebbleGettingStoreLength, err)
+	}
+	return count, nil
```

Comment:
It might be worth adding a `len uint64` field to the `pebbleKVStore` struct and like in a linked list keeping track of the length through the `Set` and `Delete` and `ClearAll` methods etc so we don't have to error here or iterate.

Reply:
We could do that in the future if there are performance issues on relay miner. I don't think we use `Len()` anywhere though, I was tempted to delete it as we don't need it.

Reply:
Len in used internally in the SMT IIRC otherwise it wouldn't be in the interface. I'd have to check where though

Reply:
@h5law just checked - `Len()` is only used in tests. I wouldn't spend time on optimizing this call then.